modules/default/updatenotification/node_helper.js

```diff
@@ -54,34 +56,78 @@ module.exports = NodeHelper.create({
-		let remotes = await git.getRemotes(true);
+	execShell: async function (command) {
+		let res = { stdout: "", stderr: "" };
+		const { stdout, stderr } = await exec(command);
 
-		if (remotes.length < 1 || remotes[0].name.length < 1) {
-			throw new Error("No valid remote for folder " + moduleFolder);
+		res.stdout = stdout;
+		res.stderr = stderr;
+		return res;
+	},
+
+	isGitRepo: async function (moduleFolder) {
+		let res = await this.execShell("cd " + moduleFolder + " && git remote -v");
+		if (res.stderr) {
+			Log.error("Failed to fetch git data for " + moduleFolder + ": " + res.stderr);
+			return false;
+		} else {
+			return true;
 		}
+	},
 
-		return git;
+	getRepoInfo: async function (repo) {
+		let gitInfo = {
+			module: repo.module,
+			// commits behind:
+			behind: 0,
+			// branch name:
+			current: "",
+			// current hash:
+			hash: "",
+			// remote branch:
+			tracking: ""
+		};
+		let res = await this.execShell("cd " + repo.folder + " && git rev-parse HEAD");
+		if (res.stderr) {
+			Log.error("Failed to get current commit hash for " + repo.module + ": " + res.stderr);
+		}
+		gitInfo.hash = res.stdout;
+		res = await this.execShell("cd " + repo.folder + " && git status -sb");
+		if (res.stderr) {
+			Log.error("Failed to get git status for " + repo.module + ": " + res.stderr);
+			// exit without git status info
+			return;
+		}
+		// get branch and remote
+		let status = res.stdout.split("\n")[0];
+		status = status.match(/(?![.#])([^.]*)/g);
+		gitInfo.current = status[0].trim();
+		status = status[1].split(" ");
+		gitInfo.tracking = status[0].trim();
+		if (status[2]) {
+			// git fetch was already called before so `git status -sb` delivers already the behind number
+			gitInfo.behind = parseInt(status[2].substring(0, status[2].length - 1));
+			return gitInfo;
+		}
+		res = await this.execShell("cd " + repo.folder + " && git fetch --dry-run");
+		// here the result is in stderr
+		if (res.stderr === "") return;
+		// set default > 0
+		gitInfo.behind = 1;
+		let refs = res.stderr.match(/s*([a-z,0-9]+[.][.][a-z,0-9]+)s*/g)[0];
+		if (refs === "") {
+			return gitInfo;
+		}
+		res = await this.execShell("cd " + repo.folder + " && git rev-list --ancestry-path --count " + refs);
+		gitInfo.behind = parseInt(res.stdout);
+		return gitInfo;
 	},
 
 	performFetch: async function () {
-		for (let sg of simpleGits) {
-			try {
-				let fetchData = await sg.git.fetch(["--dry-run"]).status();
-				let logData = await sg.git.log({ "-1": null });
-
-				if (logData.latest && "hash" in logData.latest) {
-					this.sendSocketNotification("STATUS", {
-						module: sg.module,
-						behind: fetchData.behind,
-						current: fetchData.current,
-						hash: logData.latest.hash,
-						tracking: fetchData.tracking
-					});
-				}
-			} catch (err) {
-				Log.error("Failed to fetch git data for " + sg.module + ": " + err);
+		for (let repo of gitRepos) {
+			const gitInfo = await this.getRepoInfo(repo);
```

Comment:
I'm wondering if that git status message can be mocked so that we could write some tests....

Reply:
not sure what you are talking about here, the `git status -sb` output (see new comments) or the `gitInfo` object.

Reply:
Sorry for my late night mumbling: Just ment to ask, if it is possible in your opinion to write unit tests for your new code?

Reply:
will put it on my list (not in this PR).